### PR TITLE
test: replace metrics collection and calibrate resource utilization thresholds for perf tests

### PR DESCRIPTION
**Pull request by AndrewMitchell25**
Fixes #N/A 

### Description
 Replaces the pprof-based resource utilization collection in performance tests with direct scraping of Karpenter's `/metrics` endpoint. Calibrates new thresholds using `process_resident_memory_bytes` and `process_cpu_seconds_total` across 10 CI runs.

 ### What changed

  | | Upstream (current) | This PR |
  |---|---|---|
  | **Metric source** | pprof `/debug/pprof/heap` and `/debug/pprof/profile?seconds=20` | Karpenter's `/metrics` endpoint (`process_resident_memory_bytes`, `process_cpu_seconds_total`) |
  | **Memory metric** | `inuse_space` from pprof (Go heap allocations only) | `process_resident_memory_bytes` (full RSS — what the OOM killer uses) |
  | **CPU metric** | Total nanoseconds from a 20s CPU profile sample | Rate computed from `process_cpu_seconds_total` counter deltas between 5s polls |
  | **Collection method** | New port-forward + pprof fetch every 5s (hardcoded port 1024) | API server pod proxy (`ProxyGet`) — no port-forward needed |
  | **Metrics parsing** | pprof binary protobuf (`parseProfileValue`) | `prometheus/common/expfmt` text parser |
  | **Statistics** | Peak only (`KarpenterMemoryMB`, `KarpenterCPUNanos`) | Time series with P95/Avg/Max via `montanaflynn/stats` |
  | **Assertions** | Peak memory < threshold, peak CPU nanos < threshold | P95 memory < threshold, Avg CPU < threshold |
  | **Goroutine safety** | Uses `Expect*` helpers (panics in background goroutines) | Uses error-returning helpers safe for background goroutines |
  | **Resource limits** | 1 CPU / 1 GiB (causes throttling on heavy tests) | 2 CPU / 2 GiB (allows bursting without throttling) |

### Why

  The upstream pprof approach has reliability issues in GitHub Actions CI:

  1. **`inuse_space` ≠ RSS**: pprof reports Go heap allocations only. RSS includes stack, mmap, runtime overhead, and fragmentation. A pod can pass the heap threshold but still get OOM-killed. RSS is what container memory limits enforce.

  2. **CPU nanos from pprof is noisy**: pprof CPU profiles sample at 100Hz for 20 seconds. The total nanoseconds depend on how many goroutines are active during that specific window, which varies with GHA runner scheduling. This makes results inconsistent across runs.

  3. **Peak-only is fragile**: A single GC spike or scheduling burst dominates the metric. P95 over a time series is more robust — it characterizes sustained behavior rather than transient spikes.

  4. **Port-forward churn**: Creating a new port-forward every 5 seconds (with a hardcoded port) causes connection failures and races in Kind clusters on resource-constrained runners.

  5. **1 CPU limit causes throttling**: At 1 core limit, CFS bandwidth throttling freezes all goroutines (including the metrics/pprof handler) during peak scheduling, making the endpoint unresponsive.

### Threshold methodology

  Thresholds are calculated using `mean + 3.76σ` to target a **0.1% combined flake rate** across all ~12 assertions per suite run:

  P(any flake) = 1 - (1-p)^12 = 0.001
  → p ≈ 0.0000833 per test
  → 3.76 sigma (one-sided)

  <details>
  <summary>Calibration data: P95 Memory (MB) across 10 runs</summary>

  | Test | Phase | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Run 6 | Run 7 | Run 8 | Run 9 | Run 10 | Max | Threshold |
  |------|-------|-------|-------|-------|-------|--------|--------|--------|--------|--------|--------|-----|-----------|
  | Basic Scale | scale-out | 228.75 | 218.48 | 244.96 | 223.43 | 236.81 | 232.42 | 235.49 | 241.01 | 233.25 | 235.18 | 244.96 | 260 |
  | Basic Scale | consolidation | 248.89 | 249.15 | 254.79 | 245.52 | 248.86 | 252.29 | 248.64 | 251.02 | 251.17 | 243.43 | 254.79 | 260 |
  | Drift | initial | 305.56 | 334.76 | 335.60 | 373.34 | 337.66 | 352.54 | 354.60 | 368.19 | 313.16 | 341.47 | 373.34 | 425 |
  | Drift | drift | 458.38 | 452.84 | 453.13 | 444.06 | 440.96 | 447.56 | 449.44 | 454.81 | 453.69 | 449.34 | 458.38 | 470 |
  | Wide | scale-out | 282.64 | 306.42 | 269.77 | 294.03 | 307.85 | 279.93 | 326.35 | 325.40 | 292.43 | 305.90 | 326.35 | 370 |
  | Wide | consolidation | 281.57 | 298.98 | 320.38 | 287.41 | 293.21 | 282.56 | 290.70 | 293.80 | 281.43 | 310.27 | 320.38 | 340 |
  | Host Name Spreading | scale-out | 563.87 | 556.67 | 582.45 | 583.49 | 558.28 | 478.27 | 571.94 | 561.02 | 503.77 | 566.02 | 583.49 | 680 |
  | Host Name Spreading | consolidation | 492.75 | 505.53 | 500.45 | 551.00 | 523.06 | 504.02 | 490.55 | 508.10 | 498.55 | 494.48 | 551.00 | 575 |
  | Host Name Spreading XL | scale-out | 1062.82 | 1217.37 | 1218.88 | 1161.07 | 1125.04 | 1173.22 | 1171.38 | 1166.57 | 1177.19 | 1133.98 | 1218.88 | 1330 |
  | Host Name Spreading XL | consolidation | 1013.99 | 999.33 | 1044.13 | 983.46 | 966.19 | 1140.80 | 1020.11 | 1019.77 | 1023.89 | 1016.19 | 1140.80 | 1200 |
  | Self Anti-Affinity | scale-out | 519.92 | 506.53 | 549.65 | 477.58 | 503.29 | 503.21 | 548.17 | 480.10 | 495.06 | 555.20 | 555.20 | 620 |
  | Self Anti-Affinity | interference | 948.66 | 934.19 | 942.82 | 696.05 | 869.30 | 896.24 | 878.15 | 763.18 | 846.07 | 978.72 | 978.72 | 1205 |
  | Self Anti-Affinity | consolidation | 798.41 | 770.21 | 776.62 | 783.48 | 779.43 | 794.16 | 771.89 | 764.01 | 783.96 | 786.17 | 798.41 | 820 |
  | Do Not Disrupt | scale-out | 542.38 | 553.54 | 581.03 | 538.66 | 579.14 | 533.77 | 566.52 | — | 552.94 | 602.29 | 602.29 | 645 |
  | Do Not Disrupt | consolidation | 476.29 | 454.17 | 459.09 | 454.38 | 519.84 | 514.28 | 482.04 | — | 506.05 | 501.61 | 519.84 | 585 |

  </details>

  <details>
  <summary>Calibration data: Avg CPU (cores) across 10 runs</summary>

  | Test | Phase | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Run 6 | Run 7 | Run 8 | Run 9 | Run 10 | Max | Threshold |
  |------|-------|-------|-------|-------|-------|--------|--------|--------|--------|--------|--------|-----|-----------|
  | Basic Scale | scale-out | 0.39 | 0.32 | 0.42 | 0.32 | 0.39 | 0.41 | 0.39 | 0.42 | 0.39 | 0.39 | 0.42 | 0.50 |
  | Basic Scale | consolidation | 0.16 | 0.15 | 0.19 | 0.14 | 0.19 | 0.20 | 0.15 | 0.20 | 0.17 | 0.16 | 0.20 | 0.25 |
  | Drift | initial | 0.56 | 0.54 | 0.57 | 0.65 | 0.56 | 0.56 | 0.60 | 0.58 | 0.56 | 0.63 | 0.65 | 0.70 |
  | Drift | drift | 0.87 | 0.85 | 0.68 | 0.79 | 0.80 | 0.82 | 0.85 | 0.77 | 0.87 | 0.76 | 0.87 | 1.05 |
  | Wide | scale-out | 0.56 | 0.55 | 0.48 | 0.48 | 0.44 | 0.53 | 0.48 | 0.54 | 0.51 | 0.59 | 0.59 | 0.70 |
  | Wide | consolidation | 0.23 | 0.24 | 0.25 | 0.25 | 0.22 | 0.21 | 0.23 | 0.22 | 0.24 | 0.22 | 0.25 | 0.30 |
  | Host Name Spreading | scale-out | 0.86 | 0.69 | 0.94 | 0.75 | 0.70 | 0.89 | 0.80 | 0.88 | 0.79 | 0.93 | 0.94 | 1.15 |
  | Host Name Spreading | consolidation | 0.58 | 0.78 | 0.54 | 0.63 | 0.80 | 0.59 | 0.64 | 0.56 | 0.60 | 0.53 | 0.80 | 1.00 |
  | Host Name Spreading XL | scale-out | 1.03 | 1.27 | 1.19 | 1.15 | 1.13 | 1.16 | 1.05 | 1.15 | 1.10 | 1.22 | 1.27 | 1.40 |
  | Host Name Spreading XL | consolidation | 1.28 | 1.34 | 1.07 | 0.98 | 1.34 | 1.29 | 1.37 | 1.24 | 1.21 | 1.16 | 1.37 | 1.70 |
  | Self Anti-Affinity | scale-out | 0.76 | 0.62 | 0.64 | 0.70 | 0.80 | 0.68 | 0.74 | 0.66 | 0.69 | 0.70 | 0.80 | 0.90 |
  | Self Anti-Affinity | interference | 0.92 | 1.05 | 0.87 | 0.74 | 0.93 | 1.05 | 0.97 | 1.03 | 0.85 | 0.96 | 1.05 | 1.30 |
  | Self Anti-Affinity | consolidation | 0.66 | 0.65 | 0.58 | 0.69 | 0.57 | 0.57 | 0.63 | 0.57 | 0.59 | 0.62 | 0.69 | 0.80 |
  | Do Not Disrupt | scale-out | 0.77 | 0.89 | 0.84 | 0.75 | 0.67 | 0.73 | 0.90 | — | 0.65 | 0.73 | 0.90 | 1.10 |
  | Do Not Disrupt | consolidation | 0.57 | 0.31 | 0.38 | 0.49 | 0.36 | 0.52 | 0.40 | — | 0.56 | 0.50 | 0.57 | 0.80 |

  </details>

  <details>
  <summary>Threshold math: P95 Memory (MB)</summary>

  | Test | Phase | Mean | StdDev | CV% | Threshold (mean + 3.76σ) | Multiplier |
  |------|-------|------|--------|-----|--------------------------|------------|
  | Basic Scale | scale-out | 233.0 | 7.8 | 3.4% | 260 | 1.13x |
  | Basic Scale | consolidation | 249.4 | 3.2 | 1.3% | 260 | 1.05x |
  | Drift | initial | 341.7 | 21.6 | 6.3% | 425 | 1.24x |
  | Drift | drift | 450.4 | 5.2 | 1.2% | 470 | 1.04x |
  | Wide | scale-out | 299.1 | 18.8 | 6.3% | 370 | 1.24x |
  | Wide | consolidation | 294.0 | 12.8 | 4.4% | 340 | 1.16x |
  | Host Name Spreading | scale-out | 552.6 | 34.2 | 6.2% | 680 | 1.23x |
  | Host Name Spreading | consolidation | 506.8 | 18.1 | 3.6% | 575 | 1.13x |
  | Host Name Spreading XL | scale-out | 1160.8 | 45.6 | 3.9% | 1330 | 1.15x |
  | Host Name Spreading XL | consolidation | 1022.8 | 46.9 | 4.6% | 1200 | 1.17x |
  | Self Anti-Affinity | scale-out | 513.9 | 28.5 | 5.5% | 620 | 1.21x |
  | Self Anti-Affinity | interference | 875.3 | 88.3 | 10.1% | 1205 | 1.38x |
  | Self Anti-Affinity | consolidation | 780.8 | 10.7 | 1.4% | 820 | 1.05x |
  | Do Not Disrupt | scale-out | 561.1 | 22.8 | 4.1% | 645 | 1.15x |
  | Do Not Disrupt | consolidation | 485.3 | 26.1 | 5.4% | 585 | 1.20x |

  </details>

  <details>
  <summary>Threshold math: Avg CPU (cores)</summary>

  | Test | Phase | Mean | StdDev | CV% | Threshold (mean + 3.76σ) | Multiplier |
  |------|-------|------|--------|-----|--------------------------|------------|
  | Basic Scale | scale-out | 0.38 | 0.036 | 9.4% | 0.50 | 1.35x |
  | Basic Scale | consolidation | 0.17 | 0.021 | 12.1% | 0.25 | 1.46x |
  | Drift | initial | 0.58 | 0.035 | 6.0% | 0.70 | 1.23x |
  | Drift | drift | 0.81 | 0.061 | 7.6% | 1.05 | 1.28x |
  | Wide | scale-out | 0.52 | 0.045 | 8.7% | 0.70 | 1.33x |
  | Wide | consolidation | 0.23 | 0.013 | 5.6% | 0.30 | 1.21x |
  | Host Name Spreading | scale-out | 0.82 | 0.090 | 11.0% | 1.15 | 1.41x |
  | Host Name Spreading | consolidation | 0.63 | 0.094 | 15.0% | 1.00 | 1.56x |
  | Host Name Spreading XL | scale-out | 1.15 | 0.072 | 6.3% | 1.40 | 1.24x |
  | Host Name Spreading XL | consolidation | 1.23 | 0.127 | 10.3% | 1.70 | 1.39x |
  | Self Anti-Affinity | scale-out | 0.70 | 0.056 | 8.0% | 0.90 | 1.30x |
  | Self Anti-Affinity | interference | 0.94 | 0.096 | 10.3% | 1.30 | 1.39x |
  | Self Anti-Affinity | consolidation | 0.61 | 0.044 | 7.2% | 0.80 | 1.27x |
  | Do Not Disrupt | scale-out | 0.77 | 0.088 | 11.5% | 1.10 | 1.43x |
  | Do Not Disrupt | consolidation | 0.45 | 0.095 | 20.8% | 0.80 | 1.78x |

  </details>

  ### How was this change tested?
  10 runs of the perf test suite passing with new collection/thresholds - [Example](https://github.com/AndrewMitchell25/karpenter/actions/runs/26599840752)

  Most recent run - https://github.com/AndrewMitchell25/karpenter/actions/runs/27295543739

  By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nathangeology** [comment]
I like this approach much better than what we were doing for characterizing Karpenter's controller resource usage. 10 runs isn't a lot to characterize the distribution and variation of the results, but based on the numbers and thresholds you've posted you'd expect to have a flake in about 1 of 50 CI runs based on the threshold you've chosen. Keep in mind the distinction between the statistics you are collecting during the test vs the distribution of the summary aggregates you are collecting for each run. You set your pass/fail threshold on an aggregate stat that has some variation from one run to the next. By chance this will sometimes exceed your chosen regression threshold even though there is no regression. You run 7 tests per canary run or GA run so you'd expect about 1 of 8 of your total CI runs to fail due to chance (not a regression) if the distribution is normal (which might be approximately normal due to CLT since you are looking at a percentile stat for your aggregate threshold). I think you'd be better off recording the data in the report and tracking the data across runs instead of trying to decide if a regression has occurred in a single test.

**Review by nathangeology — COMMENTED** [review]
A few things I noticed while reading through:

**1. Silent clamping of negative CPU rate**

In `recordSample()`:
```go
if cpuRate < 0 {
    cpuRate = 0
}
```

A negative rate means `process_cpu_seconds_total` went backward — that's a pod restart or counter reset. Clamping to zero silently records a false 0-core sample that pollutes the P95. Better to detect the reset, log it, and reset `prevCPUSeconds` so the next sample is clean.

**2. Goroutine lifecycle in `PortForwardPod`**

```go
go func() {
    <-ctx.Done()
    close(stop)
}()
```

If `PortForwardPod` returns early via `errCh` (port-forward fails immediately), this goroutine stays alive until the context is canceled. If the context lives longer than the caller expects, `ForwardPorts` may hang. Worth either documenting the caller's responsibility to cancel ctx in all paths, or adding explicit cleanup.

**3. CPU filtering logic in `computeStats`**

```go
for _, s := range samples {
    if s.CPUCores > 0 || len(cpuValues) > 0 {
        cpuValues = append(cpuValues, s.CPUCores)
    }
}
```

The intent is to skip the first sample (always 0 because rate needs two points). But this condition skips all leading zeros, not just the first sample. If there are multiple initial zeros (e.g., scrape started before the controller did real work), they'd all be skipped. And if a legitimate zero-CPU sample appears later, it's included. Simpler to just `samples[1:]` explicitly.

**Review by nathangeology — COMMENTED** [review]
A few things I noticed while reading through.

**Comment by nathangeology** [comment]
Re: my comments above — I see these are already addressed in the latest commit. Nice work on the counter-reset detection and the explicit `samples[1:]` fix. The port-forward goroutine lifecycle point may still apply (haven't rechecked that path closely).

File: test/pkg/environment/common/karpenter_metrics_poller.go [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+	firstSample    bool
+	sampleNum      int
+}
+
+func (mp *KarpenterMetricsPoller) run(ctx context.Context) {
+	defer close(mp.done)
+
+	state := &pollerState{
+		localPort:   rand.IntnRange(10000, 49151),
+		firstSample: true,
```
Comment (nathangeology) [inline test/pkg/environment/common/karpenter_metrics_poller.go]: Both the profiler and metrics poller pick random ports from the same range at near-simultaneous startup. Collision probability is low (~1/39000) but nonzero. Have you considered using port 0 for OS-assigned ephemeral ports, or offsetting the ranges?

File: test/pkg/environment/common/karpenter_metrics_poller.go [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+	state.localPort = rand.IntnRange(10000, 49151)
+	state.pfCtx, state.pfCancel = context.WithCancel(ctx)
+	if pfErr := mp.establishPortForward(state.pfCtx, state.localPort); pfErr != nil {
+		mp.recordError(fmt.Errorf("port-forward re-establish failed: %w", pfErr))
+		GinkgoWriter.Printf("KarpenterMetricsPoller: reconnect failed, giving up: %v\n", pfErr)
+		state.pfCancel()
+		return false
+	}
+	GinkgoWriter.Printf("KarpenterMetricsPoller: reconnected successfully on port %d\n", state.localPort)
+	return true
```
Comment (nathangeology) [inline test/pkg/environment/common/karpenter_metrics_poller.go]: After reconnecting, `prevCPUSeconds` and `prevTime` still hold values from before the gap. The next CPU rate will be diluted by the reconnection downtime. Should we reset `state.firstSample = true` here so the next scrape establishes a clean baseline?

File: test/pkg/environment/common/karpenter_metrics_poller.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+// The first sample's CPU value (always 0) is excluded from CPU statistics.
+func computeStats(samples []ResourceSample) ResourceStats {
+	if len(samples) == 0 {
+		return ResourceStats{}
+	}
+
+	memValues := make([]float64, len(samples))
+	var memSum, memMax float64
+
+	for i, s := range samples {
```
Comment (nathangeology) [inline test/pkg/environment/common/karpenter_metrics_poller.go:263]: `computeStats` and `percentile` are pure functions with edge cases (0 samples, 1 sample, gaps from counter resets). A small table-driven unit test would catch regressions and document the first-sample exclusion logic.

**Review by nathangeology — COMMENTED** [review]
Some additional observations on the current code.

File: test/pkg/environment/common/karpenter_metrics_poller.go (hunk 1 above)
Comment (nathangeology) [reply test/pkg/environment/common/karpenter_metrics_poller.go]: (Pretty minor, but could lead to a wierd bug some time haha)

File: test/pkg/environment/common/karpenter_metrics_poller.go (hunk 2 above)
Comment (nathangeology) [reply test/pkg/environment/common/karpenter_metrics_poller.go]: (Not sure on this one)

File: test/pkg/environment/common/karpenter_metrics_poller.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	if len(cpuValues) > 0 {
+		stats.P95CPUCores = percentile(cpuValues, 0.95)
+		stats.AvgCPUCores = cpuSum / float64(len(cpuValues))
+		stats.MaxCPUCores = cpuMax
+	}
+
+	return stats
+}
+
+func percentile(values []float64, p float64) float64 {
```
Comment (DerekFrank) [inline test/pkg/environment/common/karpenter_metrics_poller.go]: Should we be using something like montanaflynn/stats for this?

Comment (AndrewMitchell25) [reply test/pkg/environment/common/karpenter_metrics_poller.go]: Updated

**Review by DerekFrank — APPROVED** [review]
/lgtm
